### ISSUE TITLE
Create branches (and/or Issues) automatically and only open PR's when tests are passing

### ISSUE THREAD

**Issue by bglusman** [dependabot/feedback#507]

**Comment by greysteil** [comment]
We thought about this a while ago, and decided that creating PRs immediately was the best approach for a bunch of reasons:
- You want to know about it when an update breaks your tests (even if you may not want to merge it)
- Some folks have CI that only runs on PRs, not branches
- Functionality like `@dependabot merge`, which merges after tests, makes only creating PRs after CI has run less relevant

We might revisit in future, or make this an option, but for now we're happy with the current approach. Do you have a use case for only creating PRs after tests pass?

**Comment by bglusman** [comment]
Hmm, well, the use case is we're busy and mostly focused on having Dependabot help us out by making the easy stuff easier... an option would be nice, or an option that, after opening PR, if it fails, closing and/or reopening as an issue that basically says merging this update will require human intervention and not just merging based on a quick read of changes/versions/gut feel... but I can see your points and makes sense as a default for sure, but right now it would lead to a bit less noise if it mostly opened/left opened only passing PRs....

**Comment by greysteil** [comment]
Got you. I think a "convert to issue" flow might be a good option here - it's not uncommon for folks to want to keep their PR queue very small, so having a quick way to convert a failing PR to an issue would help. That way you'd also be able to see what tests failed if/when you decide to tackle the issue. Would that work?

**Comment by bglusman** [comment]
Yeah, so do you mean a new action `@dependabot move to issue` or do you mean a config option to automatically move failing PR's to issues?  Either would be nice, though I'd slightly prefer the latter, right now I'm doing `@dependabot ignore this patch` as a way to defer thinking about for now (though a handful initially I just did `@dependabot ignore this dependency` and a secondary issue might be now I don't see ignored stuff listed in any central place on dependabot app page, would be nice to have a place to see currently configured ignores and/or who ignored them when/allow optional annotation on why)

**Comment by jwalton** [comment]
I actually want almost exactly the opposite of this; I want Dependabot to not create a PR or an issue at all unless there's something for me to do.

Greenkeeper takes almost the opposite approach to this, and while I find dependabot overall a lot more polished, this is the one thing I really like about Greenkeeper.  (I mostly work in node.js, so I'll use JS as an example); when a new patch version of lodash comes out, GK will create a branch, update lodash, check to see if tests are passing, optionally update my lockfile with the new lodash in master, and it's done.  There's no issue, there's no PR, it just handles it for me in the background.  If tests fail, GK opens an issue for me to draw my attention to it.  If it's an out-of-range update, GK will open a PR.

Over the past couple of weeks I've upgraded all the dependencies on about six of our private repos and turned on dependabot on them.  We have auto-merge enabled, which at least prevents me from having to manually merge PRs, but that doesn't change the fact that my inbox (and all of my co-workers inboxes) look like this now:

![Screen Shot 2019-06-14 at 9 56 54 AM](https://user-images.githubusercontent.com/1771003/59514659-5047ec80-8e8b-11e9-9537-d6a8cf0a5714.png)

All my teammates are mad at me.  :P  I'm certainly not going to turn dependabot on for any more projects until we figure out how to fix this.  We're considering whether or not we should just delete all our lockfiles just to shut up dependabot.

**Comment by greysteil** [comment]
This is a really interesting one - basically I think you want a very different approach for libraries (where you want the Greenkeeper style approach) to applications (where you want the Dependabot approach).

We love what the folks at Greenkeeper built, so never wanted to tread on their toes for the library use case, but now we're part of GitHub it probably makes sense for us to allow Dependabot to be configured for the above for folks that want it.

We're in a crunch time right now with almost no resource so we can't take on new feature work right now, but I'll keep this is mind.

**Comment by jwalton** [comment]
Thanks for the insight.  :)  I think it goes a little deeper than application vs. library though.  A PR generates email notifications, so a PR should be something actionable - "Hey, you have something you need to do/a decision to make".  If Dependabot is going to auto-merge the PR for me as soon as it passes, though, there's nothing for me to do, so the PR should probably never have been raised in the first place.

It's definitely something you want to be able to turn off/on though, as you point out a lot of people only run tests on PRs (we run a subset of tests on commit and only the full suite on a PR, although at least in Travis it's pretty trivial to turn on full tests for dependabot branches because they have such nice well formatted names).
